hera_cal/smooth_cal.py

```diff
@@ -105,9 +106,109 @@ def time_filter(gains, wgts, times, filter_scale=1800.0, nMirrors=0):
+    
+    # Build fourier space image and kernel for deconvolution
+    window = aipy.dsp.gen_window(len(freqs), window=window, **win_kwargs)
+    image = np.fft.ifft2(gains * rephasor * wgts * window)
+    kernel = np.fft.ifft2(wgts * window)
+    
+    # set up "area", the set of Fourier modes that are allowed to be non-zero in the CLEAN
+    if filter_mode == 'rect':
+        area = np.outer(np.where(np.abs(fringes) < fringe_scale, 1, 0), 
+                        np.where(np.abs(delays) < delay_scale, 1, 0))
+    elif filter_mode == 'plus':
+        area = np.zeros(image.shape, dtype=int)
+        area[0] = np.where(np.abs(delays) < delay_scale, 1, 0)
+        area[:, 0] = np.where(np.abs(fringes) < fringe_scale, 1, 0)
+    else:
+        raise ValueError("CLEAN mode {} not recognized. Must be 'rect' or 'plus'.".format(filter_mode))
+
+    # perform deconvolution
+    CLEAN, info = aipy.deconv.clean(image, kernel, tol=tol, area=area, stop_if_div=False, maxiter=maxiter)
+    filtered = np.fft.fft2(CLEAN + info['res'] * area)
+    del info['res']  # this matches the convention in uvtools.dspec.high_pass_fourier_filter
+    return filtered / rephasor, info
+
+
+def pick_reference_antenna(flags):
+    '''Pick a refrence antenna that has the minimum number of per-antenna flags.
+    
+    Arguments:
+        flags: dictionary mapping antenna keys to flag waterfall where True means flagged
+
+    Returns:
+        refant: key of the antenna with the minimum number of flags. Tie goes to the first in a sorted list of keys.
+    '''
+    flags_per_ant = {ant: np.sum(f) for ant, f in flags.items()}
+    refant = sorted([ant for ant, nflags in flags_per_ant.items() if nflags == np.min(flags_per_ant.values())])[0]
+    return refant
+
+
+def rephase_to_refant(gains, refant, flags=None):
+    '''Rephase all gains in place so that the phase of the reference antenna is 0 for all times and frequencies.
+
+    Arguments:
+        gains: Dictionary mapping antenna keys to gain waterfalls. Modified in place.
+        refant: Antenna key of antenna to make the reference antenna
+        flags: Optional dictionary mapping antenna keys to flag waterfall.
+            Used only to verify that all gains are flagged where the refant is flagged.
+    '''
+    refant_phasor = gains[refant] / np.abs(gains[refant])
+    for ant in gains.keys():
+        if flags is not None and np.any(flags[refant][np.logical_not(flags[ant])]):
+            raise ValueError('The chosen reference antenna', refant, 'is flagged in at least one place where antenna',
+                             ant, 'is not, so automatic reference antenna selection has failed.')
+        else:
+            gains[ant] = gains[ant] / refant_phasor
+
+
 class CalibrationSmoother():
 
-    def __init__(self, calfits_list, flags_npz_list=[], antflag_thresh=0.0):
+    def __init__(self, calfits_list, flags_npz_list=[], pick_refant=False, antflag_thresh=0.0):
```

Comment:
Please add `pick_refant` to the docstring.

Reply:
Also, should there be an option to manually select the refant?

Reply:
Done. 

For manual refant selection, I'd prefer to punt on this. I'm not sure if we need it and how we implement it depends on whether we want to specify different reference antennas for each polarization or not. If we need it, we can add it pretty straightforwardly.